src/ConfCrypt/Template.hs

```diff
@@ -0,0 +1,70 @@
+module ConfCrypt.Template (
+    renderTemplate
+    ) where
+
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Data.Text
+import ConfCrypt.Types (Parameter(..))
+import qualified Data.Text as T
+
+-- "text %k=%v %%"
+-- "text foo=bar %"
+-- ["text ", k, "=", v, " %"]
+-- %[a-z] %x %y %z
+
+
+renderTemplate :: Text -> Parameter -> Text
+renderTemplate template param = case parse parseTemplate "" template of
+    Left err -> pack . show $ err
+    Right parsed -> foldMap replaceVars parsed
+    where
+        replaceVars (Text_ t)         = t
+        replaceVars (Variable_ Name)  = paramName param
+        replaceVars (Variable_ Value) = paramValue param
+        replaceVars (Variable_ Type)  = T.pack . show . paramType $ param
+
+
+type Parser = Parsec MyParseError Text
+type ApplyTemplate = Text -> Text -> Text -> Text
+newtype MyParseError = MyParseError Text deriving (Show, Ord, Eq)
+
+-- type Parser = Parsec ParseError Text
+
+parseTemplate :: Parser [Template]
+parseTemplate =
+    many (txt <|> var)
+    where
+        var = Variable_ <$> parseVariable
+        txt = Text_ <$> parseLiteral
+
+data Template = Variable_ Variable | Text_ Text
+    deriving Show
+
+parseLiteral :: Parser Text
+parseLiteral = let
+    escapedPercent = do
+        _ <- try $ string' "%%"
+        pure "%"
+    otherText = pack <$> some (noneOf ['%'])
+    in escapedPercent <|> otherText
+
+
+data Variable = Name | Value | Type deriving Show
+
+parseVariable :: Parser Variable
+parseVariable = let
+    tryName = do
```

Comment:
This is an ideal place for the half-apply operator [`*>`](https://www.stackage.org/haddock/lts-13.6/base-4.12.0.0/Prelude.html#v:-42--62-). It's a version of regular apply operator `<*>`, except it ignores the left side (and there is also its evil twin `<*`, which ignores the right side), so it can be used to do a thing, ignore its result, and then return another thing:

```
tryName = try (string' "%n") *> pure Name
```

And then, `*> pure` is equivalent to [`$>`](https://www.stackage.org/haddock/lts-13.6/base-4.12.0.0/Data-Functor.html#v:-36--62-) from `Data.Functor`:

```
tryName = (string' "%n") $> Name
```

Though personally I find its evil twin more readable:

```
tryName = Name <$ try (string' "%n")
```